hub/src/server/revocations.js

```diff
@@ -0,0 +1,147 @@
+    evictionLogTimeout.unref()
+  }
+
+  handleCacheEvictions() {
+    if (this.currentCacheEvictions > 0) {
+      logger.warn(`Gaia authentication token timestamp cache evicted ${this.currentCacheEvictions} entries in the last 10 minutes. Consider increasing 'authTimestampCacheSize'.`)
+      this.currentCacheEvictions = 0
+    }
+  }
+
+  getAuthTimestampFileDir(bucketAddress: string) {
+    return `${bucketAddress}-auth`
+  }
+
+  async readAuthTimestamp(bucketAddress: string): Promise<number> {
+
+    const authTimestampDir = this.getAuthTimestampFileDir(bucketAddress)
+    
+    let fetchResponse
+    let authNumberText
+    try {
+      const authNumberFileUrl = `${this.readUrlPrefix}${authTimestampDir}/${AUTH_TIMESTAMP_FILE_NAME}`
+      fetchResponse = await fetch(authNumberFileUrl, {
+        redirect: 'manual'
+      })
+      authNumberText = await fetchResponse.text()
+    } catch (err) {
+      // Catch any errors that may occur from network issues during `fetch` and `.text()` async operations..
+      const errMsg = (err instanceof Error) ? err.message : err
+      throw new errors.ValidationError(`Error trying to fetch bucket authentication revocation timestamp: ${errMsg}`)
+    }
+
+    if (fetchResponse.ok) {
+      const authNumber = parseInt(authNumberText)
+      if (Number.isFinite(authNumber)) {
+        return authNumber
+      } else {
+        throw new errors.ValidationError(`Bucket contained an invalid authentication revocation timestamp: ${authNumberText}`)
+      }
+    } else if (fetchResponse.status === 404) {
+      // 404 incidates no revocation file has been created.
+      return 0
+    } else {
+      throw new errors.ValidationError(`Error trying to fetch bucket authentication revocation timestamp: server returned ${fetchResponse.status} - ${fetchResponse.statusText}`)
+    }
+
+  }
+
+  async getAuthTimestamp(bucketAddress: string): Promise<number> {
+    // First perform fast check if auth number exists in cache..
+    let authTimestamp = this.cache.get(bucketAddress)
+    if (authTimestamp) {
+      return authTimestamp
+    }
+
+    // Nothing in cache, perform slower driver read.
+    authTimestamp = await this.readAuthTimestamp(bucketAddress)
+
+    // Recheck cache for a larger timestamp to avoid race conditions from slow storage.
+    const cachedTimestamp = this.cache.get(bucketAddress)
+    if (cachedTimestamp && cachedTimestamp > authTimestamp) {
+      authTimestamp = cachedTimestamp
+    }
+
+    // Cache result for fast lookup later.
+    this.cache.set(bucketAddress, authTimestamp)
+
+    return authTimestamp
+  }
+
+  async writeAuthTimestamp(bucketAddress: string, timestamp: number) : Promise<void> {
+
+    // Recheck cache for a larger timestamp to avoid race conditions from slow storage.
+    const cachedTimestamp = this.cache.get(bucketAddress)
+    if (cachedTimestamp && cachedTimestamp > timestamp) {
```

Comment:
The check here, as well as the `this.cache.set()` call, need to happen _after_ the call to `this.driver.performWrite()`.

Reply:
This should be fixed now.